### ISSUE TITLE
Download Jobs seems to run in infinite loop.

### ISSUE THREAD

**Issue by cdietrich** [eclipse-m2e/m2e-core#331]
On my workspace the DownloadJob runs all the time. 
it seems to schedule itself via

```
DownloadSourcesJob$DownloadRequest.<init>(IProject, IPackageFragmentRoot, ArtifactKey, boolean, boolean) line: 79	
DownloadSourcesJob.addDownloadRequest(IProject, IPackageFragmentRoot, ArtifactKey, boolean, boolean) line: 352	
DownloadSourcesJob.scheduleDownload(IProject, IPackageFragmentRoot, ArtifactKey, boolean, boolean) line: 342	
DownloadSourcesJob.scheduleDownload(IProject, ArtifactKey, boolean, boolean) line: 362	
BuildPathManager.scheduleDownload(IProject, MavenProject, ArtifactKey, boolean, boolean) line: 919	
BuildPathManager.configureAttachedSourcesAndJavadoc(IMavenProjectFacade, Properties, ClasspathDescriptor, IProgressMonitor) line: 343	
BuildPathManager.getClasspath(IMavenProjectFacade, int, Properties, boolean, IProgressMonitor) line: 270	
BuildPathManager.getClasspath(IProject, int, boolean, IProgressMonitor) line: 413	
BuildPathManager.getClasspath(IProject, int, IProgressMonitor) line: 393	
BuildPathManager.getClasspath(IProject, IProgressMonitor) line: 421	
BuildPathManager.updateClasspath(IProject, IProgressMonitor) line: 206	
DownloadSourcesJob.updateClasspath(BuildPathManager, Set<IProject>, Map<IPackageFragmentRoot,Attachments>, IProgressMonitor) line: 207	
DownloadSourcesJob.run(IProgressMonitor) line: 185	
```
the problematic place may be

```
  private void scheduleDownload(IProject project, MavenProject mavenProject, ArtifactKey artifact,
      boolean downloadSources, boolean downloadJavadoc) throws CoreException {
    ArtifactKey[] attached = getAttachedSourcesAndJavadoc(artifact, mavenProject.getRemoteArtifactRepositories(),
        downloadSources, downloadJavadoc);

    if(attached[0] != null || attached[1] != null) {
      downloadSourcesJob.scheduleDownload(project, artifact, downloadSources, downloadJavadoc);
    }
  }
```
the null checks dont also check the downloadSources and downloadJavadoc flags
(attached = `[com.google.guava:guava:30.1-jre:sources, null]`) in my case
`downloadSources	true`
`downloadJavadoc	false`

```
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<groupId>xxxxxx</groupId>
	<artifactId>a22222</artifactId>
	<version>0.0.1-SNAPSHOT</version>

	<dependencies>
		<dependency>
			<groupId>com.google.guava</groupId>
			<artifactId>guava</artifactId>
			<version>30.1-jre</version>
			<type>jar</type>
		</dependency>
	</dependencies>
</project>
```

**Comment by cdietrich** [comment]
https://user-images.githubusercontent.com/238736/134147657-e1cbb75f-8cb0-4cc5-917a-10b64663531a.mov

**Comment by cdietrich** [comment]
[eclipse-config.txt](https://github.com/eclipse-m2e/m2e-core/files/7202519/eclipse-config.txt)

**Comment by cdietrich** [comment]
downloadMaven: why this

```
if(attached[1] != null) {
      try {
        javadoc = download(attached[1], repositories, monitor);
        log.info("Downloaded javadoc for " + artifact.toString());
      } catch(CoreException e) {
        log.error("Could not download javadoc for " + artifact.toString(), e); //$NON-NLS-1$
      }
    }
```
isnt the check the wrong way?

**Comment by HannesWell** [comment]
This seems to be a duplicate of bug #252 that is resolved already, but was not yet released.
> 
> 
> [eclipse-config.txt](https://github.com/eclipse-m2e/m2e-core/files/7202519/eclipse-config.txt)

Form this config I read that you use the latest release m2e 1.18.1. But the fix of the mentioned issue was not yet released.
Please try the latest m2e snapshot to see if this resolves your problem. You can use the following repository:
https://download.eclipse.org/technology/m2e/snapshots/1.18.2/latest/

**Comment by cdietrich** [comment]
i found that one problem might have been that the sources was older than the jar

```
cdietrich:~/git/m2e-core[tags/1.18.1] > ls -laT  /Users/cdietrich/.m2/repository/com/google/guava/guava/30.1-jre/guava-30.1-jre-sources.jar
-rw-r--r--  1 cdietrich  staff  1705726 25 Jan 12:42:25 2021 /Users/cdietrich/.m2/repository/com/google/guava/guava/30.1-jre/guava-30.1-jre-sources.jar
cdietrich:~/git/m2e-core[tags/1.18.1] > ls -laT /Users/cdietrich/.m2/repository/com/google/guava/guava/30.1-jre/guava-30.1-jre.jar
-rw-r--r--  1 cdietrich  staff  2862361 25 Jan 12:42:30 2021 /Users/cdietrich/.m2/repository/com/google/guava/guava/30.1-jre/guava-30.1-jre.jar
```

deleting the sources solved the problem.
will also check the new snapshot

**Comment by HannesWell** [comment]
> 
> 
> i found that one problem might have been that the sources was older than the jar
> 
> ```
> cdietrich:~/git/m2e-core[tags/1.18.1] > ls -laT  /Users/cdietrich/.m2/repository/com/google/guava/guava/30.1-jre/guava-30.1-jre-sources.jar
> -rw-r--r--  1 cdietrich  staff  1705726 25 Jan 12:42:25 2021 /Users/cdietrich/.m2/repository/com/google/guava/guava/30.1-jre/guava-30.1-jre-sources.jar
> cdietrich:~/git/m2e-core[tags/1.18.1] > ls -laT /Users/cdietrich/.m2/repository/com/google/guava/guava/30.1-jre/guava-30.1-jre.jar
> -rw-r--r--  1 cdietrich  staff  2862361 25 Jan 12:42:30 2021 /Users/cdietrich/.m2/repository/com/google/guava/guava/30.1-jre/guava-30.1-jre.jar
> ```
> 
> deleting the sources solved the problem.
> will also check the new snapshot

I faced the exact same problem for a SNAPSHOT version and fixed it (this case is also explicitly covered by tests now).
Yes, deleting the sources is a workaround.

**Comment by cdietrich** [comment]
i can confirm that the new snapshot also solves the problem

**Comment by HannesWell** [comment]
> 
> 
> i can confirm that the new snapshot also solves the problem

Great. Then this issue can be closed.

**Comment by HannesWell** [comment]
Alright. Can you provide a (minimalistic) reproducer?
Your initial example worked well for me.

**Comment by cdietrich** [comment]
i fear i cant. the Xtext developement env is complex. 
i still have not understood where the cycle should be broken now.
can you give me some hints where to look in code when remote debugging

**Comment by HannesWell** [comment]
> 
> 
> i fear i cant. the Xtext developement env is complex.
> i still have not understood where the cycle should be broken now.
> can you give me some hints where to look in code when remote debugging

OK. Understand that.
I have described my findings here: https://github.com/eclipse-m2e/m2e-core/issues/252#issuecomment-903173237

Therefore I recommend to put a break-point at `DownloadSourcesJob.scheduleDownload(IProject, IPackageFragmentRoot, ArtifactKey, boolean, boolean)`, propably best inside the body of the if-statement `if(requests.add(request)) {`, line 327.

The request set is intended to break cycles when `BuildPathManager.updateClasspath` is called in the end of a DownloadSourcesJob run.


> 
> e.g. i dont understand this logic
> 
> ```
>   private void scheduleDownload(IProject project, MavenProject mavenProject, ArtifactKey artifact,
>       boolean downloadSources, boolean downloadJavadoc) throws CoreException {
>     ArtifactKey[] attached = getAttachedSourcesAndJavadoc(artifact, mavenProject.getRemoteArtifactRepositories(),
>         downloadSources, downloadJavadoc);
> 
>     if(attached[0] != null || attached[1] != null) {
>       downloadSourcesJob.scheduleDownload(project, artifact, downloadSources, downloadJavadoc);
>     }
>   }
> ```

The name `getAttachedSourcesAndJavadoc()` is misleading IMHO. As far as I have understand the logic this method returns the artifacts to download respectively to attach, not those that are already attached.

**Comment by cdietrich** [comment]
ok. is it possible the sources and code in the snapshot update site dont match

for the stacktrace

```
DownloadSourcesJob.scheduleDownload(IProject, IPackageFragmentRoot, ArtifactKey, boolean, boolean) line: 323	
DownloadSourcesJob.scheduleDownload(IProject, ArtifactKey, boolean, boolean) line: 340	
BuildPathManager.scheduleDownload(IProject, MavenProject, ArtifactKey, boolean, boolean) line: 867	
BuildPathManager.configureAttachedSourcesAndJavadoc(IMavenProjectFacade, Properties, ClasspathDescriptor, IProgressMonitor) line: 323	
BuildPathManager.getClasspath(IMavenProjectFacade, int, Properties, boolean, IProgressMonitor) line: 250	
BuildPathManager.getClasspath(IProject, int, boolean, IProgressMonitor) line: 372	
BuildPathManager.getClasspath(IProject, int, IProgressMonitor) line: 354	
BuildPathManager.getClasspath(IProject, IProgressMonitor) line: 380	
BuildPathManager.updateClasspath(IProject, IProgressMonitor) line: 208	
DownloadSourcesJob.updateClasspath(BuildPathManager, Set<IProject>, Map<IPackageFragmentRoot,Attachments>, IProgressMonitor) line: 199	
DownloadSourcesJob.run(IProgressMonitor) line: 178	
```
the variables view in configureAttachedSourcesAndJavadoc shows 
no `downloadJavaDoc`
variable

```
        if(aKey != null) { // maybe we should try to find artifactKey little harder here?
          boolean isSnapshot = aKey.getVersion().endsWith("-SNAPSHOT");
          // We should update a sources/javadoc jar for a snapshot in case they're already downloaded.
          File mainFile = desc.getPath() != null ? desc.getPath().toFile() : null;
          File srcFile = srcPath != null ? srcPath.toFile() : null;
          boolean downloadSources = (srcPath == null && mavenConfiguration.isDownloadSources())
              || (isSnapshot && isLastModifiedBefore(srcFile, mainFile));
          File javaDocFile = javaDocUrl != null ? getAttachedArtifactFile(aKey, CLASSIFIER_JAVADOC) : null;
          boolean downloadJavaDoc = (javaDocUrl == null && mavenConfiguration.isDownloadJavaDoc())
              || (isSnapshot && isLastModifiedBefore(javaDocFile, mainFile));
          scheduleDownload(facade.getProject(), facade.getMavenProject(monitor), aKey, downloadSources,
              downloadJavaDoc);
        }
```

**Comment by cdietrich** [comment]
shouldnt the 
 if(requests.add(request)) {
 be a containsKey?

**Comment by cdietrich** [comment]
![Bildschirmfoto 2021-09-21 um 16 06 41](https://user-images.githubusercontent.com/238736/134185906-f291e8b0-8b9d-4abd-a4ea-4b078f7ef1b5.png)


here the project is flagged

**Comment by cdietrich** [comment]
problem is again timestamps

```
ls -laT ~/.m2/repository/org/eclipse/xtext/org.eclipse.xtext.common.types/2.26.0-SNAPSHOT
total 3272
drwxr-xr-x  17 cdietrich  staff     544 21 Sep 15:12:35 2021 .
drwxr-xr-x  50 cdietrich  staff    1600 21 Sep 14:20:44 2021 ..
-rw-r--r--   1 cdietrich  staff     410 21 Sep 14:32:56 2021 _remote.repositories
-rw-r--r--   1 cdietrich  staff    1099 21 Sep 16:11:13 2021 m2e-lastUpdated.properties
-rw-r--r--   1 cdietrich  staff    1407 21 Sep 15:12:34 2021 maven-metadata-xtext-extras-from-jenkins.xml
-rw-r--r--   1 cdietrich  staff      40 21 Sep 15:12:35 2021 maven-metadata-xtext-extras-from-jenkins.xml.sha1
-rw-r--r--   1 cdietrich  staff  335511 21 Sep 14:20:45 2021 org.eclipse.xtext.common.types-2.26.0-20210918.073411-1-sources.jar
-rw-r--r--   1 cdietrich  staff      40 21 Sep 14:20:46 2021 org.eclipse.xtext.common.types-2.26.0-20210918.073411-1-sources.jar.sha1
-rw-r--r--   1 cdietrich  staff  471839 21 Sep 14:32:56 2021 org.eclipse.xtext.common.types-2.26.0-20210918.073411-1.jar
-rw-r--r--   1 cdietrich  staff      40 21 Sep 14:32:56 2021 org.eclipse.xtext.common.types-2.26.0-20210918.073411-1.jar.sha1
-rw-r--r--   1 cdietrich  staff    7016 21 Sep 14:32:46 2021 org.eclipse.xtext.common.types-2.26.0-20210918.073411-1.pom
-rw-r--r--   1 cdietrich  staff    1222 21 Sep 14:32:46 2021 org.eclipse.xtext.common.types-2.26.0-20210918.073411-1.pom.lastUpdated
-rw-r--r--   1 cdietrich  staff      40 21 Sep 14:32:46 2021 org.eclipse.xtext.common.types-2.26.0-20210918.073411-1.pom.sha1
-rw-r--r--   1 cdietrich  staff  335511 21 Sep 14:20:45 2021 org.eclipse.xtext.common.types-2.26.0-SNAPSHOT-sources.jar
-rw-r--r--   1 cdietrich  staff  471839 21 Sep 14:32:56 2021 org.eclipse.xtext.common.types-2.26.0-SNAPSHOT.jar
-rw-r--r--   1 cdietrich  staff    7016 21 Sep 14:32:46 2021 org.eclipse.xtext.common.types-2.26.0-SNAPSHOT.pom
-rw-r--r--   1 cdietrich  staff     650 21 Sep 15:12:35 2021 resolver-status.properties
```

in this case the source seems to be older than the jar again.
am not sure why this is
(can come from https://ci.eclipse.org/xtext/job/xtext-extras/job/master/lastSuccessfulBuild/artifact/build/maven-repository/ and https://oss.sonatype.org/content/repositories/snapshots)

**Comment by cdietrich** [comment]
reproducer

```
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<groupId>xxxxxx</groupId>
	<artifactId>a22222</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	
	<repositories>
		<repository>
			<id>oss</id>
			<url>https://oss.sonatype.org/content/repositories/snapshots</url>
		</repository>
	</repositories>

	<dependencies>
		<dependency>
			<groupId>com.google.guava</groupId>
			<artifactId>guava</artifactId>
			<version>30.1-jre</version>
			<type>jar</type>
		</dependency>
		<dependency>
			<groupId>org.eclipse.xtext</groupId>
			<artifactId>org.eclipse.xtext.common.types</artifactId>
			<version>2.26.0-SNAPSHOT</version>
		</dependency>
	</dependencies>
</project>
```

might be needed to have older javadoc than jar once to trigger

**Comment by cdietrich** [comment]
even more minimal

```
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<groupId>xxxxxx</groupId>
	<artifactId>a22222</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	
	<repositories>
		<repository>
			<id>oss</id>
			<url>https://oss.sonatype.org/content/repositories/snapshots</url>
		</repository>
	</repositories>

	<dependencies>
		<dependency>
			<groupId>com.google.guava</groupId>
			<artifactId>guava</artifactId>
			<version>30.1-jre</version>
			<type>jar</type>
		</dependency>
		<dependency>
			<groupId>org.eclipse.xtext</groupId>
			<artifactId>org.eclipse.xtext.xbase.lib</artifactId>
			<version>2.26.0-SNAPSHOT</version>
		</dependency>
	</dependencies>
</project>
```

```
~/.m2/repository/org/eclipse/xtext/org.eclipse.xtext.xbase.lib/2.26.0-SNAPSHOT > ls -la
total 3696
drwxr-xr-x  30 cdietrich  staff     960 21 Sep 16:42 .
drwxr-xr-x  97 cdietrich  staff    3104 21 Sep 14:20 ..
-rw-r--r--   1 cdietrich  staff     753 21 Sep 16:42 _remote.repositories
-rw-r--r--   1 cdietrich  staff    2188 21 Sep 16:46 m2e-lastUpdated.properties
-rw-r--r--   1 cdietrich  staff    2039 21 Sep 16:42 maven-metadata-oss.xml
-rw-r--r--   1 cdietrich  staff      40 21 Sep 16:42 maven-metadata-oss.xml.sha1
-rw-r--r--   1 cdietrich  staff    1404 21 Sep 16:19 maven-metadata-xtext-lib-from-jenkins.xml
-rw-r--r--   1 cdietrich  staff      40 21 Sep 16:19 maven-metadata-xtext-lib-from-jenkins.xml.sha1
-rw-r--r--   1 cdietrich  staff  327200 21 Sep 16:32 org.eclipse.xtext.xbase.lib-2.26.0-20210918.072404-1-javadoc.jar
-rw-r--r--   1 cdietrich  staff      40 21 Sep 16:32 org.eclipse.xtext.xbase.lib-2.26.0-20210918.072404-1-javadoc.jar.sha1
-rw-r--r--   1 cdietrich  staff   93178 21 Sep 14:20 org.eclipse.xtext.xbase.lib-2.26.0-20210918.072404-1-sources.jar
-rw-r--r--   1 cdietrich  staff      40 21 Sep 14:20 org.eclipse.xtext.xbase.lib-2.26.0-20210918.072404-1-sources.jar.sha1
-rw-r--r--   1 cdietrich  staff  146245 21 Sep 14:32 org.eclipse.xtext.xbase.lib-2.26.0-20210918.072404-1.jar
-rw-r--r--   1 cdietrich  staff      40 21 Sep 14:32 org.eclipse.xtext.xbase.lib-2.26.0-20210918.072404-1.jar.sha1
-rw-r--r--   1 cdietrich  staff    5374 21 Sep 14:32 org.eclipse.xtext.xbase.lib-2.26.0-20210918.072404-1.pom
-rw-r--r--   1 cdietrich  staff     735 21 Sep 14:32 org.eclipse.xtext.xbase.lib-2.26.0-20210918.072404-1.pom.lastUpdated
-rw-r--r--   1 cdietrich  staff      40 21 Sep 14:32 org.eclipse.xtext.xbase.lib-2.26.0-20210918.072404-1.pom.sha1
-rw-r--r--   1 cdietrich  staff  327200 21 Sep 16:42 org.eclipse.xtext.xbase.lib-2.26.0-20210921.015111-239-javadoc.jar
-rw-r--r--   1 cdietrich  staff      40 21 Sep 16:42 org.eclipse.xtext.xbase.lib-2.26.0-20210921.015111-239-javadoc.jar.sha1
-rw-r--r--   1 cdietrich  staff  104208 21 Sep 16:42 org.eclipse.xtext.xbase.lib-2.26.0-20210921.015111-239-sources.jar
-rw-r--r--   1 cdietrich  staff      40 21 Sep 16:42 org.eclipse.xtext.xbase.lib-2.26.0-20210921.015111-239-sources.jar.sha1
-rw-r--r--   1 cdietrich  staff  164399 21 Sep 16:42 org.eclipse.xtext.xbase.lib-2.26.0-20210921.015111-239.jar
-rw-r--r--   1 cdietrich  staff      40 21 Sep 16:42 org.eclipse.xtext.xbase.lib-2.26.0-20210921.015111-239.jar.sha1
-rw-r--r--   1 cdietrich  staff    5374 21 Sep 16:42 org.eclipse.xtext.xbase.lib-2.26.0-20210921.015111-239.pom
-rw-r--r--   1 cdietrich  staff      40 21 Sep 16:42 org.eclipse.xtext.xbase.lib-2.26.0-20210921.015111-239.pom.sha1
-rw-r--r--   1 cdietrich  staff  327200 21 Sep 16:42 org.eclipse.xtext.xbase.lib-2.26.0-SNAPSHOT-javadoc.jar
-rw-r--r--   1 cdietrich  staff  104208 21 Sep 16:42 org.eclipse.xtext.xbase.lib-2.26.0-SNAPSHOT-sources.jar
-rw-r--r--   1 cdietrich  staff  164399 21 Sep 16:42 org.eclipse.xtext.xbase.lib-2.26.0-SNAPSHOT.jar
-rw-r--r--   1 cdietrich  staff    5374 21 Sep 16:42 org.eclipse.xtext.xbase.lib-2.26.0-SNAPSHOT.pom
-rw-r--r--   1 cdietrich  staff     702 21 Sep 16:42 resolver-status.properties
```

**Comment by cdietrich** [comment]
ok, seems to be a combination of these projects

**Comment by HannesWell** [comment]
Thanks for the reproducer. I will check them this evening.

Just to be clear: I can create a new Maven-project in a clean workspace with one of the given reproducer poms? Or do I have to checkout the xtext projects you mentioned previously?

**Comment by cdietrich** [comment]
two new simple maven projects with no archetype selections
![Bildschirmfoto 2021-09-21 um 17 06 39](https://user-images.githubusercontent.com/238736/134196683-936ef0af-ce89-46b3-868b-90f630cee24a.png)

**Comment by HannesWell** [comment]
> 
> 
> ok. is it possible the sources and code in the snapshot update site dont match
> 
This would surprise me very much and would be a major malfunction of the m2e build pipeline. What I think is more likely is that the sources in the debugging IDE do not match the plug-in version in the debugged Eclipse instance, one of them might not be up-to-date?

**Comment by cdietrich** [comment]
yes version is up to date
i get this only after a wild combination of deletes of the jars (not the sources)
maven update project calls and maven download sources calls on the set of projects. 

i still cannot relate this to a consistent behaviour

**Comment by cdietrich** [comment]
ok i found another way to trigger it.
when i select a jar in the container and call download sources on it i get the loop too
![Bildschirmfoto 2021-09-21 um 21 18 05](https://user-images.githubusercontent.com/238736/134233952-4f7a6be3-c446-49af-bdcd-b3451e207e4d.png)

**Comment by HannesWell** [comment]
OK, got it. I can reproduce it.
The PR is on the way.

**Comment by HannesWell** [comment]
When looking again into the sources of the `DownloadSourcesJob` I just found the root of the problem as a gone-wrong clean up of myself (I suspect I falsely resolved some merge conflicts), even before I could reproduce the problem.
First I thought the scenario in which that flaw would be harmful was mostly theoretical, but after I was able to reproduce your case I noticed that I was wrong and you hit exactly that case. With the change in the PR the `DownloadSourcesJob` terminated immediately so it should fix your problem.
My apologies for the inconvenience.

**Comment by cdietrich** [comment]
looks better now. will done some more testing this evening

**Comment by cdietrich** [comment]
more tests were all successful. can no longer reproduce with latest 1.18.2. thx

**Comment by HannesWell** [comment]
Great, your welcome and thanks for providing the reproducer.

**Comment by gkfirst8** [comment]
I have now noticed the same behavior again in Eclipse:
Version: 2022-09 (4.25.0)
Build id: 20220908-1902

Hope this can somehow be fixed again as this makes working in Eclipse as slow as typing in vim ;-)
